### PR TITLE
Support Apache Commons Daemon methods in XmlConfiguration

### PR DESCRIPTION
**Pull request by gregw**
This is a minimal change to allow XmlConfiguration to be substituted for start.Main if using a --dry-run to avoid an exec.

### CODE REVIEW

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 1]
```diff
@@ -361,6 +377,8 @@ else if (__factoryLoader != null)
      */
     public Object configure(Object obj) throws Exception
     {
+        if (_objects != null)
+            throw new IllegalStateException();
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: Add a reason to the ISE

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 2]
```diff
@@ -375,6 +393,8 @@ public Object configure(Object obj) throws Exception
      */
     public Object configure() throws Exception
     {
+        if (_objects != null)
+            throw new IllegalStateException();
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: Add a reason to the ISE

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 3]
```diff
@@ -1833,24 +1853,8 @@ Args asEmptyVarArgs(Class<?> varArgType)
...
-     * <p>
-     * Any IDs created in a configuration are passed to the next configuration file on the command line using {@link #getIdMap()}.
-     * This allows objects with IDs created in one config file to be referenced in subsequent config files on the command line.
-     *
-     * @param args array of property and xml configuration filenames or {@link Resource}s.
-     * @throws Exception if the XML configurations cannot be run
-     */
-    public static void main(final String... args) throws Exception
+    // implement Apache commons daemon (jsvc) lifecycle methods (init, start, stop, destroy)
+    public void init(String[] args) throws Exception
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: Add `if (objects == null) throw ISE`?

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 4]
```diff
@@ -1919,6 +1911,82 @@ else if (arg.toLowerCase(Locale.ENGLISH).endsWith(".properties"))
         }
     }
 
+    // implement Apache commons daemon (jsvc) lifecycle methods (init, start, stop, destroy)
+    public void start() throws Exception
+    {
+        if (_objects == null)
+            throw new IllegalStateException();
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: Add reason to ISE.

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 5]
```diff
@@ -1919,6 +1911,82 @@ else if (arg.toLowerCase(Locale.ENGLISH).endsWith(".properties"))
...
+            }
+        }
+
+    }
+
+    // implement Apache commons daemon (jsvc) lifecycle methods (init, start, stop, destroy)
+    public void stop() throws Exception
+    {
+        if (_objects == null)
+            throw new IllegalStateException();
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: Add a reason

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 6]
```diff
@@ -1919,6 +1911,82 @@ else if (arg.toLowerCase(Locale.ENGLISH).endsWith(".properties"))
...
+
+        // For all objects created by XmlConfigurations, start them if they are lifecycles.
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: s/start/stop

File: jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java [hunk 7]
```diff
@@ -1919,6 +1911,82 @@ else if (arg.toLowerCase(Locale.ENGLISH).endsWith(".properties"))
...
+                    lc.stop();
+            }
+        }
+    }
+
+    // implement Apache commons daemon (jsvc) lifecycle methods (init, start, stop, destroy)
+    public void destroy()
+    {
+        if (_objects == null)
+            throw new IllegalStateException();
```
Comment (janbartel) [inline jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java]: Add reason

**Review by janbartel — APPROVED** [review]
The approach seems promising to me: small scope of changes but satisfies the requirements. If we can't think of a usecase this doesn't cover, I'm in favour of proceeding with it (but see some review comments).

**Comment by gregw** [comment]
@joakime you've convinced me that we need both approaches (methods on XmlConfiguration, but also a --no-exec flag).   I will have a go at implementing that in this PR as well.

**Comment by gregw** [comment]
@joakime @janbartel @sbordet Where are we on this one?
I think the file args feature of the JVM that simone pointed out should fix the issue - although it still requires the output of --dry-run to be edited (to remove the java command itself).

So I think the changes in this PR are still relevant:
 + The --dry-run=parts stuff makes it easier to generate partial command lines.   This could perhaps be augmented with Joakims suggestion of shell/cmd style output.
 + The --no-exec change is kind of interesting, if not strictly necessary.
 + The extra methods on XmlConfiguration should fulfil our contract that dry-run is interchangeable with start.jar... but again not strictly necessary.

So currently I'm 80% thinking dial this back to just --dry-run=parts..... but 15% of me says keep all if it in 10.... 5% says all of it in 9.4

thoughts???

**Comment by joakime** [comment]
We need an issue filed for this one.

Otherwise it wont show up in the VERSION.txt

**Comment by gregw** [comment]
@joakime oh I didn't know..... well let's work out what we are going to change.... if anything... and then create an issue.

**Review by sbordet — COMMENTED** [review]
@gregw `java -jar start.jar --dry-run` will output the command line with the same `java` you used to issue the command, so you can do `/path/to/fancy/java -jar start.jar --dry-run` and you will get it right.

If that is true, I think we should go with this, i.e. probably nothing to do.
I would wait for another use case before adding `--dry-run=parts`.

**Comment by joakime** [comment]
Having the `--dry-run=parts` is at least useful for separating out the java vm from --dry-run (which is present currently).
So that you can build a resulting `java @jvmrunfile` style execution.

**Comment by sbordet** [comment]
@joakime but why we need to do `java @file`? can't we just do `/path/to/fancy/java -jar start.jar --dry-run > exec.sh` and then just `exec.sh`?
Why we need to split `java` out of the rest?

**Comment by joakime** [comment]
> @joakime but why we need to do `java @file`? can't we just do `/path/to/fancy/java -jar start.jar --dry-run > exec.sh` and then just `exec.sh`?
> Why we need to split `java` out of the rest?

That wont work on Windows.
The Command line would be too large.

There's a few paths this can take, and a few use cases.

1. Using Apache Commons Daemon.
2. Starting Jetty Standalone from a script without a forked JVM.
3. Integrating Jetty Standalone into part of a larger script ecosystem. (eg: Docker)

Your proposal would result in ...

1. Working for Unix, but not Windows.
2. Working for Unix, but not Windows.
3. Complex post-dry-run parsing of the output from --dry-run to integrate into a larger java / jvm setup and execution. (perhaps even with multiple calls at different times in the docker image hierarchy)

**Comment by sbordet** [comment]
So the only thing that is needed is `--dry-run=parts` but not every single part?

**Comment by gregw** [comment]
@sbordet @joakime I've removed all but the parts and added to the documentation to match.

**Pull request merged**